types/jsreport-core/index.d.ts

```diff
@@ -0,0 +1,98 @@
+// Type definitions for jsreport-core 1.5
+// Project: http://jsreport.net
+// Definitions by: taoqf <https://github.com/taoqf>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+// TypeScript Version: 2.6
+
+/// <reference types="node" />
+
+import { Buffer } from 'buffer';
+import { Readable } from 'stream';
+
+type helpers = (...args: any[]) => any;
+
+interface RenderOptions {
+	template: {
+		content: string;
+		engine: 'jsrender' | 'handlebars' | 'ejs' | 'jade' | string;
+		helpers?: string | { [fun: string]: helpers };
+		recipe: 'phantom-pdf' | 'electron-pdf' | 'text' | 'xlsx' | 'html-to-xlsx' | 'phantom-image' | 'html-to-text' | 'fop-pdf' | 'client-html' | 'wrapped-html' | 'wkhtmltopdf' | string;
+	};
+	data?: any;
+}
+
+interface Report {
+	content: Buffer;
+	stream: Readable;
+	headers: {
+		[header: string]: string | number | boolean;
+	};
+}
+
+interface Request {
+	template: {
+		content: string;
+	};
+}
+
+// interface Response {
+// 	// todo
+// }
+
+type Response = any;
+
+interface Listener {
+	add(type: string, callback: (req: Request, res: Response, err: any) => void): void;
+}
+
+interface Logger {
+	add(logger: any, options?: {
+		level: 'debug' | 'info' | 'log' | 'warn' | 'error';
+	}): void;
+}
+
+interface Collection {
+	find(query: {
+		[field: string]: any;
+	}): Promise<any>;
+}
+
+interface DocumentStore {
+	collection(options: string): Collection;
+}
+
+interface JsReporter {
+	afterRenderListeners: Listener;
+	afterTemplatingEnginesExecutedListeners: Listener;
+	beforeRenderListeners: Listener;
+	documentStore: DocumentStore;
+	initializeListeners: Listener;
+	logger: Logger;
+	validateRenderListeners: Listener;
+	init(): Promise<void>;
+	render(options: RenderOptions): Promise<Report>;
+	use(extension: any): any;
+}
+
+type JsReportStatic = (options?: Partial<{
+	autoTempCleanup: boolean;
+	dataDirectory: string;
+	extensionsLocationCache: boolean;
+	loadConfig: boolean;
+	logger: {
+		silent: boolean;
+	};
+	rootDirectory: string;
+	scripts: {
+		allowedModules: string[];
+	};
+	tasks: {
+		[task: string]: any;
+	};
+	tempDirectory: string;
+}>) => JsReporter;
+
+declare const JsReport: JsReportStatic;
```

Comment:
`declare function JsReport(options?: { autoTempCleanup?: boolean; ... }): JsReporter;`

Reply:
After all, I've got this:
```sh
➜  DefinitelyTyped git:(master) cnpm run lint jsreport-core

> definitely-typed@0.0.1 lint /home/taoqf/git/DefinitelyTyped
> dtslint types "jsreport-core"

Test with 2.6
Test with 2.5
Test with 2.4
Test with 2.3
Test with 2.2
Error: /home/taoqf/git/DefinitelyTyped/types/jsreport-core/index.d.ts:1:1
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'RegExp'.
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'Array'.
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'Function'.
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'IArguments'.
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'Number'.
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'Object'.
ERROR: 1:1    expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find global type 'Boolean'.
ERROR: 11:43  expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
A rest parameter must be of an array type.
ERROR: 56:6   expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find name 'Promise'.
ERROR: 71:10  expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find name 'Promise'.
ERROR: 72:34  expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find name 'Promise'.
ERROR: 76:37  expect  Compile error in typescript@2.2 but not in typescript@2.3.
Fix with a comment '// TypeScript Version: 2.3' just under the header.
Cannot find name 'Partial'.

/home/taoqf/git/DefinitelyTyped/types/jsreport-core/jsreport-core-tests.ts:2:27
ERROR: 2:27   expect  TypeScript@next compile error:
Module '"/home/taoqf/git/DefinitelyTyped/types/jsreport-core/index"' resolves to a non-module entity and cannot be imported using this construct.

    at /home/taoqf/git/DefinitelyTyped/node_modules/_dtslint@0.1.2@dtslint/bin/index.js:101:19
    at Generator.next (<anonymous>)
    at fulfilled (/home/taoqf/git/DefinitelyTyped/node_modules/_dtslint@0.1.2@dtslint/bin/index.js:5:58)
    at <anonymous>
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! definitely-typed@0.0.1 lint: `dtslint types "jsreport-core"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the definitely-typed@0.0.1 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/taoqf/.npm/_logs/2018-01-24T01_21_55_067Z-debug.log
```

Reply:
The whole file:
```ts
// Type definitions for jsreport-core 1.5
// Project: http://jsreport.net
// Definitions by: taoqf <https://github.com/taoqf>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

/// <reference types="node" />

import { Buffer } from 'buffer';
import { Readable } from 'stream';

type helpers = string | { [fun: string]: (...args: any[]) => any };

interface RenderOptions {
	template: {
		content: string;
		engine: 'jsrender' | 'handlebars' | 'ejs' | 'jade' | string;
		helpers?: helpers;
		recipe: 'phantom-pdf' | 'electron-pdf' | 'text' | 'xlsx' | 'html-to-xlsx' | 'phantom-image' | 'html-to-text' | 'fop-pdf' | 'client-html' | 'wrapped-html' | 'wkhtmltopdf' | string;
	};
	data?: any;
}

interface Report {
	content: Buffer;
	stream: Readable;
	headers: {
		[header: string]: string | number | boolean;
	};
}

interface Request {
	template: {
		content: string;
	};
}

// interface Response {
// 	// todo
// }

type Response = any;

interface Listener {
	add(type: string, callback: (req: Request, res: Response, err: any) => void): void;
}

interface Logger {
	add(logger: any, options?: {
		level: 'debug' | 'info' | 'log' | 'warn' | 'error';
	}): void;
}

interface Collection {
	find(query: {
		[field: string]: any;
	}): Promise<any>;
}

interface DocumentStore {
	collection(options: string): Collection;
}

interface JsReporter {
	afterRenderListeners: Listener;
	afterTemplatingEnginesExecutedListeners: Listener;
	beforeRenderListeners: Listener;
	documentStore: DocumentStore;
	initializeListeners: Listener;
	logger: Logger;
	validateRenderListeners: Listener;
	init(): Promise<void>;
	render(options: RenderOptions): Promise<Report>;
	use(extension: any): any;
}

declare function JsReport(options?: Partial<{
	autoTempCleanup: boolean;
	dataDirectory: string;
	extensionsLocationCache: boolean;
	loadConfig: boolean;
	logger: {
		silent: boolean;
	};
	rootDirectory: string;
	scripts: {
		allowedModules: string[];
	};
	tasks: {
		[task: string]: any;
	};
	tempDirectory: string;
}>): JsReporter;

export = JsReport;

```

Reply:
The `Cannot find global type` errors don't reproduce for me. Did you somehow have `"noLib": true` set in your tsconfig?
As for the `resolves to a non-module entity` error, that is because the library should be imported with `import JsReport = require('jsreport-core');` instead of `import * as JsReport from 'jsreport-core';`.

Reply:
tsconfig.json
```json
{
	"compilerOptions": {
		"module": "commonjs",
		"lib": [
			"esnext"
		],
		"strictFunctionTypes": true,
		"noImplicitAny": true,
		"noImplicitThis": true,
		"strictNullChecks": true,
		"baseUrl": "../",
		"typeRoots": [
			"../"
		],
		"types": [],
		"noEmit": true,
		"forceConsistentCasingInFileNames": true
	},
	"files": [
		"index.d.ts",
		"jsreport-core-tests.ts"
	]
}
```

Reply:
I find I could remove `// TypeScript Version: 2.3`, if i do, the errors will throw again.What can I do now? @andy-ms

Reply:
I guess if couldn't hurt to leave it in.